text/0000-mutex_traits.md

```diff
@@ -0,0 +1,174 @@
+
+A solution to this problem which would enable mutual exclusivity is to hold that the relation between types and traits is in one of three states for every type and trait, rather than two:
+
+* `T: Trait` - The type `T` impleemnts the trait `Trait`.
+* `T: !Trait` - The type `T` does not implement the trait `Trait`.
+* `T: ?Trait` - The type `T` may or may not implement the trait `Trait`.
+
+Without a contrary expression, the relationship between any `T` and any `Trait` is  `T: ?Trait`. This table documents the three relations and how they are described:
+
+|               | ?Trait            | Trait                | !Trait             |
+|---------------|-------------------|----------------------|--------------------|
+| Specific impl | by default        | impl Trait for T     | impl !Trait for T  |
+|               | impl ?Trait for T |                      |                    |
+| Default impl  | by default        | impl Trait for ..    | impl !Trait for .. |
+| Bounds        | by default        | where T: Trait       | where T: !Trait    |
+|               | where T: ?Sized   | by default for Sized |                    |
+
+## Defining ?Trait and !Trait
+
+`?Trait` and `!Trait` act as if they were marker traits. They define no methods and have no associated items. They are defined implicitly in the same scope as the definition of `Trait` and imported wherever `Trait` is imported.
+
+## Implementing ?Trait
+
+`?Trait` can only be implemented for types for which a default impl has been explicitly defined (e.g. `Send` and `Sync`). Explicit default impls of `?Trait` are not allowed. `?Trait` is implemented by default anyway, and it would not make sense to implement it except in the cases where an explicit default impl exists. As a rule, the syntax `?Trait` will be very uncommon.
+
+`?Trait` follows the same orphan rules as `Trait`.
+
+## Implementing !Trait
+
+Implementing `!Trait` for `T` is a forward-compatible guarantee that `T` does not and will not implement `Trait`. This makes negative reasoning explicit and avoids backwards compatibility hazards. It goes without saying that it would be a coherence violation for a single type to implement both `Trait` and `!Trait`.
+
+`!Trait` follows the same orphan rules as `Trait`.
+
+## Bounding by !Trait
+
+Bounding by `!Trait` requires that types _explicitly implement_ `!Trait` in order to meet that bound. As mentioned prior, this avoids the hazard that implicit negative reasoning introduces.
+
+## Syntactic sugar: Implicit `!Trait` inference
+
+If a type `T` implements a trait `Foo` which is bounded `!Bar`, an implementation of `!Bar` is inferred for `T` (unless `T` explicitly implements `!Bar`, of course). This avoids boilerplate negative impls which are inferrable from other impls for the type.
```

Comment:
I chose to propose that sugar because it seemed like there wasn't enough in the pre-RFC for folks to disagree about. :-) In seriousness, I do think this is good sugar.

It's not actually true that you have to manually implement all positive traits which are bounds, because some are implemented by default through default impls. Default impls currently can only be provided for marker traits, and of the four non-built-in traits in std::marker (`Send`, `Sync`, `Reflect`, and `Unsize`), three are implemented by default. Negative traits are all marker traits, and while inference is a different mechanism from default impls, it is similar in principle.

If a type implements a trait, it _cannot_ implement a trait which is mutually exclusive with that trait. Implementing that trait is a semantic declaration that this type does not implement traits that are mutually exclusive with the traits it does implement. Mutual exclusion implies that these traits are fundamentally disparate (e.g. `Int` and `Float` or `Num` and `Collection`), a client should not be expecting that a type which implements `Int` can be passed where a type is bound `Float`.

It seems much less of a leaping inference than the leap that Rust already makes in local contexts, which makes this code compile (and which cannot be removed).

```rust
pub trait Quux { }

pub struct Foo;
pub trait Bar { }

impl Quux for Foo { }
impl<T> Quux for T where T: Bar { }
```